modules/table/src/test/java/org/apache/ignite/internal/table/distributed/index/IndexBuilderTest.java

```diff
@@ -0,0 +1,173 @@
+import java.util.stream.IntStream;
+import org.apache.ignite.internal.replicator.ReplicaService;
+import org.apache.ignite.internal.replicator.message.ReplicaRequest;
+import org.apache.ignite.internal.storage.MvPartitionStorage;
+import org.apache.ignite.internal.storage.RowId;
+import org.apache.ignite.internal.storage.index.IndexStorage;
+import org.apache.ignite.internal.testframework.BaseIgniteAbstractTest;
+import org.apache.ignite.network.ClusterNode;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+/** For {@link IndexBuilder} testing. */
+public class IndexBuilderTest extends BaseIgniteAbstractTest {
+    private static final int TABLE_ID = 1;
+
+    private static final int INDEX_ID = 2;
+
+    private static final int PARTITION_ID = 3;
+
+    private final ReplicaService replicaService = mock(ReplicaService.class, invocation -> completedFuture(null));
+
+    private IndexBuilder indexBuilder;
+
+    @BeforeEach
+    void setUp() {
+        indexBuilder = new IndexBuilder("test", 1, replicaService);
+    }
+
+    @AfterEach
+    void tearDown() {
+        indexBuilder.close();
+    }
+
+    @Test
+    void testIndexBuildCompletionListener() {
+        CompletableFuture<Void> listenCompletionIndexBuildingFuture = listenCompletionIndexBuilding(INDEX_ID, TABLE_ID, PARTITION_ID);
+
+        scheduleBuildIndex(INDEX_ID, TABLE_ID, PARTITION_ID, List.of(rowId(PARTITION_ID)));
+
+        assertThat(listenCompletionIndexBuildingFuture, willCompleteSuccessfully());
+    }
+
+    @Test
+    void testStopListenIndexBuildCompletion() {
+        CompletableFuture<Void> invokeListenerFuture = new CompletableFuture<>();
+
+        IndexBuildCompletionListener listener = (indexId, tableId, partitionId) -> invokeListenerFuture.complete(null);
+
+        indexBuilder.listen(listener);
+        indexBuilder.stopListen(listener);
+
+        scheduleBuildIndex(INDEX_ID, TABLE_ID, PARTITION_ID, List.of(rowId(PARTITION_ID)));
+
+        assertThat(invokeListenerFuture, willTimeoutFast());
+    }
+
+    @Test
+    void testIndexBuildCompletionListenerTwoBatches() {
+        CompletableFuture<Void> listenCompletionIndexBuildingFuture = listenCompletionIndexBuilding(INDEX_ID, TABLE_ID, PARTITION_ID);
+
+        List<RowId> nextRowIdsToBuild = IntStream.range(0, 2 * IndexBuilder.BATCH_SIZE)
+                .mapToObj(i -> rowId(PARTITION_ID))
+                .collect(toList());
+
+        CompletableFuture<Void> secondInvokeReplicaServiceFuture = new CompletableFuture<>();
+
+        CompletableFuture<Void> awaitSecondInvokeForReplicaService = awaitSecondInvokeForReplicaService(secondInvokeReplicaServiceFuture);
+
+        scheduleBuildIndex(INDEX_ID, TABLE_ID, PARTITION_ID, nextRowIdsToBuild);
+
+        assertThat(awaitSecondInvokeForReplicaService, willCompleteSuccessfully());
+
+        assertFalse(listenCompletionIndexBuildingFuture.newIncompleteFuture().isDone());
```

Comment:
This looks weird as `newIncompleteFuture().isDone()` will always be `false`

Reply:
Wow, I don’t understand how this happened, I fixed it. =)